### ISSUE TITLE
insert into.values.returning(getFields).fetchOne().into(clazz) isn't mapping to 'getFields' for the clazz

### ISSUE THREAD

**Issue by VasilyShelkov** [jOOQ/jOOQ#4014]
Hi,

First time posting on here, you've got a great API by the way. Hats off to you :+1: 

I seem to be running into inconsistent behaviour where I'm creating an object which has 7 fields, it's returning only 6 of those 7 fields but is trying to map to all 7 fields of the row that are created. To be more precise it's not mapping <code>into(clazz)</code> by the fields that are passed into <code>returning(getFields)</code>.

The code I'm running is:

```java
return getDSLContext().insertInto(table, createFields)
                .values(row)
                .returning(getFields)
                .fetchOne()
                .into(clazz);
```

Where <code>table</code> is a generated table, <code>createFields</code> are the code generated tableFields (only creating 4/7 column for that specific table are being passed in, the rest have default values as set on the db), <code>row</code> is an arraylist correlating to the 4 fields being created. <code>getFields</code> are the 6 fields that I want to return. 

What happens when I've debugged is the creating and fetching are fine and it only fetches the values for the fields passed in <code>getFields</code>. However it tries to map to a constructor with all the fields for that row rather than the ones that were passed into returning().

![image](https://cloud.githubusercontent.com/assets/7474810/5990130/9bb0beca-a99a-11e4-8cff-71bd8b097406.png)

so from the image there are 6 values that are fetched by <code>fetchOne()</code> which correlate to the <code>getFields</code> that are passed into <code>.returning(getFields)</code>, however, the record's 'fields' are 7 which it tries to map to the <code>.into(clazz)</code>.

That's a lengthy explanation as to the behaviour, the reason why I believe it's inconsistent is because when I call:

```java
getDSLContext()
                .select(getAccountsFields())
                .from(ACCOUNTS)
                .where(ACCOUNTS.ACCOUNTID.equal(id))
                .fetchOne().into(Account.class);
```

where <code>getAccountFields()</code> = <code>getFields</code> in the previous example and uses the same <code>fetchOne().into(Account.class)</code> there are no mappingExceptions because it's mapping by the values returned.

As a work around I just have just got two constructors on the Account.class object, one for constructing with all the fields that <code>insertInto..</code> is returning and one for constructing with the fields that <code>select...</code> is returning. So it's not a blocker, but ideally I'd like just the one constructor as it's making the code ambiguous and inconsistent.

I hope that makes sense and if that's intended behaviour for <code>returning(getFields).fetchOne().into(clazz)</code> to map into the class differently to <code>.select(getAccountsFields().from(ACCOUNTS).where(ACCOUNTS.ACCOUNTID.equal(id)).fetchOne().into(Account.class);</code> then that's fine, I just thought I'd flag it in case it wasn't intended and is a genuine inconsistency.

I will happily add to this or answer any questions if anything was unclear. Don't very often write to github (still a growing developer !) but want to do what I can given how much I appreciate JOOQ.

Love the API regardless, and think that you've done the best at what you're trying to achieve :) have a good weekend ! PS also want to give lots of love to Lukas Eder's java 8/ general JOOQ blog posts! they're a top notch read for me.

**Comment by lukaseder** [comment]
Thanks for the awesome bug report. We'll be looking into it very soon!

**Comment by VasilyShelkov** [comment]
ok thanks for replying so rapidly :) I hope it wasn't too wordy. I haven't really submitted bugs before and am new the open source community! Any constructive feedback on that too will be appreciated too.

**Comment by lukaseder** [comment]
> Any constructive feedback on that too will be appreciated too.

Sure. Let's start with that first. I have two items:

- It's usually better to post more code and less "prosa". Code is just much easier to digest than English. In particular, I'm talking about this section:

    > Where table is a generated table, createFields are the code generated tableFields (only creating 4/7 column for that specific table are being passed in, the rest have default values as set on the db), row is an arraylist correlating to the 4 fields being created. getFields are the 6 fields that I want to return.

- Instead of inline code formatting, you could also format the code in block format. This isn't specific to GitHub, but it's particularly easy here. I'll just edit the question for you. You can see the change directly, then.
- As far as jOOQ is concerned, we usually answer questions and support requests on the [user group](https://groups.google.com/d/forum/jooq-user), and track only issues and feature requests here on GitHub. This is because more people search the user group for answers than GitHub. And we're afraid of what happens when we hit issue #6666. But if in doubt, just don't worry. We try to help wherever we can.

I'll reply to your actual questions in a separate comment.

**Comment by lukaseder** [comment]
The behaviour you're getting is "expected". The problem here is the fact that everything really looks like a single statement, and thus you concluded that the call to `into(Class)` somehow "knows" what had happened before (e.g. the call to `returning()`). But that would be incredibly hard to implement and get right, so just decouple all the calls and imagine that `returning()` will return a `Record`, and you're calling `into()` just on that single `Record`.

The full behaviour of `into()` is specified by the `RecordMapperProvider` that you are configuring. By default, this will apply the [`DefaultRecordMapper`](www.jooq.org/javadoc/latest/org/jooq/impl/DefaultRecordMapper.html)

**Comment by VasilyShelkov** [comment]
Appreciate the time you took to give the the feedback on my report. You're right I should have included the code or at least mentioned that Accout.class was a simple immutable POJO that I was mapping to. I have taken it all onboard and in the future hopefully will make me better at submitting issues or looking for support ! Maybe if I'm lucky will be able to contribute something or get involved. I can imagine it being a good workplace with you around !!

I will also be looking into the user group before coming here, especially for any support. You're right, I was trying to chain too much without the implications of it. Got a bit carried away with the functional style of writing code and not thinking about the deeper implications.

**Comment by lukaseder** [comment]
> I have taken it all onboard and in the future hopefully will make me better at submitting issues or looking for support

First step: Don't worry about it so much ;)

> Got a bit carried away with the functional style of writing code and not thinking about the deeper implications.

I guess that's a good sign, right? :)
